sogs/legacy_routes.py

```diff
@@ -416,50 +416,23 @@ def ban_checks():
     return user, room, to_ban
 
 
-def apply_ban(conn, user, room, to_ban):
-    is_mod = bool(
-        conn.execute(
-            "SELECT moderator FROM user_permissions WHERE room = ? AND user = ?",
-            (room.id, to_ban.id),
-        ).fetchone()[0]
-    )
-
-    if is_mod and not model.check_permission(user, room, admin=True):
-        app.logger.warn(
-            "Cannot ban {} from {}: the ban target is a room moderator, "
-            "but the ban initiator ({}) is not an admin".format(
-                to_ban.session_id, room.token, user.session_id
-            )
-        )
-        abort(http.FORBIDDEN)
-
-    conn.execute(
-        """
-        INSERT INTO user_permission_overrides (room, user, banned, moderator, admin)
-        VALUES (?, ?, TRUE, FALSE, FALSE)
-        ON CONFLICT (room, user) DO UPDATE SET banned = TRUE, moderator = FALSE, admin = FALSE
-        """,
-        (room.id, to_ban.id),
-    )
-
-
 @app.post("/legacy/block_list")
 def handle_legacy_ban():
     user, room, to_ban = ban_checks()
 
-    with db.conn as conn:
-        apply_ban(conn, user, room, to_ban)
+    if not model.ban_user(user, room, to_ban):
+        abort(http.FORBIDDEN)
 
     return jsonify({"status_code": 200})
 
 
 @app.post("/legacy/ban_and_delete_all")
 def handle_legacy_banhammer():
     user, room, to_ban = ban_checks()
+    if not model.ban_user(user, room, to_ban):
+        abort(http.FORBIDDEN)
 
     with db.conn as conn:
-        apply_ban(conn, user, room, to_ban)
-
```

Comment:
This is changing the behaviour here: we want to do the ban + delete in a single transaction, so that either it all happens, or none of it happens.  Now we can get a ban applied but the posts not deleted for some reason, since they aren't happening in the same transaction.

Reply:
would making this ban_user function optionally delete all posts too work?

Reply:
Feels kind of wrong to do two things at once in one method like that.

Reply:
Actually the transaction here is probably okay, in this order: if you ban but the delete fails you can try again.  (In the opposite order -- i.e. delete worked but ban didn't) it's difficult to actually repeat the ban.